doc: add naaj doc (#11208)
pingcap/docs-cn: explain-subqueries.md

```sql
create table t(a int, b int);
create table s(a int, b int);
explain select  (a, b) not in (select * from s) from t;
explain select * from t where (a, b) not in (select * from s);
```

```sql
tidb> explain select  (a, b) not in (select * from s) from t;
+-----------------------------+---------+-----------+---------------+---------------------------------------------------------------------------------------+
| id                          | estRows | task      | access object | operator info                                                                         |
+-----------------------------+---------+-----------+---------------+---------------------------------------------------------------------------------------+

> | HashJoin_8                  | 1.00    | root      |               | NAAJ anti left outer semi join, equal:[eq(test.t.b, test.s.b) eq(test.t.a, test.s.a)] |

Reviewer: NAAJ is short for `NULL AWARE ANTI JOIN`; it's redundant with the latter `anti .... semi join`.

上述查询首先读取了表 `t3`，然后根据主键开始探测 (probe) 表 `t1`。连接类型是 _anti semi join_，即反半连接：之所以使用 _anti_，是因为上述示例有不存在匹配值（即 `NOT IN`）的情况；使用 `Semi Join` 则是因为仅需要匹配第一行后就可以停止查询。

## Null-Aware Semi Join（`IN`、`= ANY` 子查询）

`IN`、`= ANY` 的集合运算符号具有特殊的三值属性（`true`、`false`、`NULL`），这意味着在其所转化得到的 Join 类型中需要对 Join key 两侧的 `NULL` 进行特殊的感知和处理。

> 对于 `IN`、`= ANY` 算子而言，其所引导的子查询，会对应地转为 Semi Join 和 Left Outer Semi Join。在上述 [Semi Join](#semi-join关联查询) 小节中， 由于其 Join key 的两侧列 `test.t1.id` 和 `test.t2.t1_id` 都是 `not NULL` 属性的，所以 Semi Join 本身不需要 Null-Aware 的性质来辅助运算。当前 TiDB 对于 Null-Aware Semi Join 没有特定的优化，其实现本质都是基于笛卡尔积加过滤 (filter) 的模式。以下为 Null-Aware 的例子：

Reviewer 2: ```suggestion
对于 `IN`、`= ANY` 算子而言，其所引导的子查询，会对应地转为 Semi Join 和 Left Outer Semi Join。在上述 [Semi Join](#semi-join关联查询) 小节中， 由于示例中 Join key 两侧的列 `test.t1.id` 和 `test.t2.t1_id` 都是 `not NULL` 属性的，所以 Semi Join 本身不需要 Null-Aware 的性质来辅助运算。当前 TiDB 对于 Null-Aware Semi Join 没有特定的优化，其实现本质都是基于笛卡尔积加过滤 (filter) 的模式。以下为 Null-Aware Semi Join 的例子：
```

| │   └─TableFullScan_12       | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                                      |
| └─HashAgg_17(Probe)          | 1.00    | root      |               | group by:test.s.a, test.s.b, funcs:firstrow(test.s.a)->test.s.a, funcs:firstrow(test.s.b)->test.s.b |
|   └─TableReader_24           | 1.00    | root      |               | data:Selection_23                                                                                   |
|     └─Selection_23           | 1.00    | cop[tikv] |               | not(isnull(test.s.a)), not(isnull(test.s.b))                                                        |
|       └─TableFullScan_22     | 1.00    | cop[tikv] | table:s       | keep order:false, stats:pseudo                                                                      |
+------------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------------+
8 rows in set (0.01 sec)
```

> 第一个查询 `EXPLAIN SELECT (a,b) IN (SELECT * FROM s) FROM t;` 中，由于 `t` 表和 `s` 表的 `a`、`b` 列都是 NULLABLE 的，所以其所转化 Left Outer Semi Join 是具有 Null-Aware 性质的。其实现是先通过笛卡尔乘积，然后将 `IN` 或者 `= ANY` 所连接的列作为普通等值条件放到 other condition 做过滤（笛卡尔积之后的 filter）实现的。

Reviewer 2: ```suggestion
第一个查询 `EXPLAIN SELECT (a,b) IN (SELECT * FROM s) FROM t;` 中，由于 `t` 表和 `s` 表的 `a`、`b` 列都是 NULLABLE 的，所以 `IN` 子查询所转化的 Left Outer Semi Join 是具有 Null-Aware 性质的。具体实现是先进行笛卡尔积，然后将 `IN` 或 `= ANY` 所连接的列作为普通等值条件放到 other condition 进行过滤（filter）。
```

> 第二个查询 `EXPLAIN SELECT * FROM t WHERE (a,b) IN (SELECT * FROM s);` 中，由于 `t` 表和 `s` 表的 `a`、`b` 列都是 NULLABLE 的，`IN` 属性本应该转为 Null-Aware 性质的 Semi Join，但当前 TiDB 有一层优化，直接将 Semi Join 转为了 Inner Join + aggregate 的方式来实现。这是因为在非 scalar 输出的 `IN` sub-query 中，`NULL` 和 `false` 具有等效结果，其下推过滤的 `NULL` 行，就其本身来说都是导致 `WHERE` 子句的否定语义，因此可以事先忽略这些行。

Reviewer 2: ```suggestion
第二个查询 `EXPLAIN SELECT * FROM t WHERE (a,b) IN (SELECT * FROM s);` 中，由于 `t` 表和 `s` 表的 `a`、`b` 列都是 NULLABLE 的，`IN` 子查询本应该转为具有 Null-Aware 性质的 Semi Join，但当前 TiDB 进行了优化，直接将 Semi Join 转为了 Inner Join + aggregate 的方式来实现。这是因为在非 scalar 输出的 `IN` 子查询中，`NULL` 和 `false` 具有等效结果，其下推过滤的 `NULL` 行导致了 `WHERE` 子句的否定语义，因此可以事先忽略这些行。
```

> **注意：**
>
> `Exists` 操作符也会转成 Semi Join，但是 `Exists` 操作符号本身不具有集合运算 Null-Aware 的性质。

## Null-Aware Anti Semi Join（`NOT IN`、`!= ALL` 子查询）

`NOT IN`、`!= ALL` 的集合运算运算具有特殊的三值属性（`true`、`false`、`NULL`），这意味着在其所转化得到的 Join 类型中需要对 Join key 两侧的 `NULL` 进行特殊的感知和处理。

> 对于 `NOT IN`、`!= ALL` 算子而言，其所引导的子查询，会对应的转为 Anti Semi Join 和 Anti Left Outer Semi Join。在上述的 [Anti Semi Join](#anti-semi-joinnot-in-子查询) 小节中，由于其 Join key 的两侧列 `test.t3.t1_id` 和 `test.t1.id` 都是 `not NULL` 属性的，所以 Anti Semi Join 本身不需要 Null-Aware 的性质来辅助计算。

Reviewer 2: ```suggestion
对于 `NOT IN`、`!= ALL` 算子而言，其所引导的子查询，会对应地转为 Anti Semi Join 和 Anti Left Outer Semi Join。在上述的 [Anti Semi Join](#anti-semi-joinnot-in-子查询) 小节中，由于示例中 Join key 两侧的列 `test.t3.t1_id` 和 `test.t1.id` 都是 `not NULL` 属性的，所以 Anti Semi Join 本身不需要 Null-Aware 的性质来辅助计算。
```

> 在 TiDB v6.3.0 及后续的版本中，TiDB 引入了针对 Null-Aware Anti Join (NAAJ) 的特殊优化：

Reviewer 3: ```suggestion
在 TiDB v6.3.0 版本，TiDB 引入了针对 Null-Aware Anti Join (NAAJ) 的如下特殊优化：
```

- 利用 Null-Aware 的等值条件 (NA-EQ) 构建哈希连接

    由于集合操作符引入的等值需要对等值两侧操作符数的 `NULL` 值做特殊处理，这里称需要 Null-Aware 的等值条件为 NA-EQ 条件。与 v6.3.0 之前版本不同的是，TiDB 不会再将 NA-EQ 条件处理成普通 EQ 条件，而是专门放置于 Join 后置的 other condition 中，匹配笛卡尔积后再判断结果集的合法性。

>     在 TiDB v6.3.0 版本中，NA-EQ 这种弱化的等值条件依然会被用来构建哈希值 (Hash Join)，大大减少了匹配时所需遍历的数据量，加速匹配过程。在 build 表 `DISTINCT()` 值很大的时候，加速效果更为显著。

Author: ```suggestion
    在 TiDB v6.3.0 版本中，NA-EQ 这种弱化的等值条件依然会被用来构建哈希值 (Hash Join)，大大减少了匹配时所需遍历的数据量，加速匹配过程。在 build 表 `DISTINCT` 值比例趋近 1 的时候，加速效果更为显著。
```

>     在 TiDB v6.3.0 及之后版本的优化中，NA-EQ 这种弱化的等值条件依然会被用来构建哈希值 (Hash Join)，大大减少了匹配时所需遍历的数据量，加速匹配过程。在 build 表 `DISTINCT()` 值很大的时候，加速效果更为显著。

Reviewer 2: ```suggestion
    在 TiDB v6.3.0 版本中，NA-EQ 这种弱化的等值条件依然会被用来构建哈希值 (Hash Join)，大大减少了匹配时所需遍历的数据量，加速匹配过程。在 build 表 `DISTINCT()` 值很大的时候，加速效果更为显著。
```

>     在 TiDB v6.3.0 及之后版本的优化中，NA-EQ 这种弱化的等值条件依然会被用来构建哈希值 (Hash Join)，大大减少了匹配时所需遍历的数据量，加速匹配过程。在 build 表 `DISTINCT()` 值很大的时候，加速效果更为显著。

Author: distinct 跟原始值有一个比例，

（1，1，1）distinct 值只有一个 （去重）= 1/3
（1，2，3）distinct 值有 3 个 =  3/3
distinct 值越大，这个比例越来越接近 1 （100%）

> - 利用两侧数据源 `NULL` 值的特殊性质加速匹配结果的返回

Author: ```suggestion
- 利用两侧数据源 `NULL` 值的特殊性质加速匹配过程的返回
```

    由于 Anti Semi Join 自身具有 CNF (Conjunctive normal form) 表达式的属性，其任何一侧出现的 `NULL` 值都会导致结果的确定性，利用这种性质可以来加速整个匹配过程的返回。

> 以下为 Null-Aware 的例子：

Reviewer 2: ```suggestion
以下为 Null-Aware Anti Semi Join 的例子：
```

>     由于 Anti Semi Join 自身具有 CNF (Conjunctive normal form) 表达式的属性，其任何一侧出现的 `NULL` 值都会导致确定的结果。利用这个性质可以来加速整个匹配过程的返回。

Reviewer 2: ```suggestion
    由于 Anti Semi Join 自身具有 CNF (Conjunctive normal form) 表达式的属性，其任何一侧出现的 `NULL` 值都会导致确定的结果。利用这个性质可以来加速整个匹配过程。
```

>     由于 Anti Semi Join 自身具有 CNF (Conjunctive normal form) 表达式的属性，其任何一侧出现的 `NULL` 值都会导致确定的结果。利用这个性质可以来加速整个匹配过程的返回。

Author: 这两个应该是一个意思，可以斟酌使用其一即可

+-----------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------+
| HashJoin_8                  | 8000.00  | root      |               | Null-aware anti semi join, equal:[eq(test.t.b, test.s.b) eq(test.t.a, test.s.a)] |
| ├─TableReader_12(Build)     | 10000.00 | root      |               | data:TableFullScan_11                                                            |
| │ └─TableFullScan_11        | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                                                   |
| └─TableReader_10(Probe)     | 10000.00 | root      |               | data:TableFullScan_9                                                             |
|   └─TableFullScan_9         | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                   |
+-----------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> 第一个查询 `EXPLAIN SELECT (a, b) NOT IN (SELECT * FROM s) FROM t;` 中，由于 `t` 表和 `s` 表的 `a`，`b` 列都是 NULLABLE 的，所以其所转化的 Left Outer Semi Join 是具有 Null-Aware 性质的。不同的是，NAAJ 优化将 NA-EQ 等值条件也作为了 Hash Join 的连接条件，大大加速的了 Join 的计算。

Reviewer 2: ```suggestion
第一个查询 `EXPLAIN SELECT (a, b) NOT IN (SELECT * FROM s) FROM t;` 中，由于 `t` 表和 `s` 表的 `a`，`b` 列都是 NULLABLE 的，所以 `NOT IN` 子查询所转化的 Left Outer Semi Join 是具有 Null-Aware 性质的。不同的是，NAAJ 优化将 NA-EQ 条件也作为了 Hash Join 的连接条件，大大加速了 Join 的计算。
```

> 第二个查询 `EXPLAIN SELECT * FROM t WHERE (a, b) NOT IN (select * FROM s);` 中，由于 `t` 表和 `s` 表的 `a`，`b` 列都是 NULLABLE 的，所以其所转化的 Anti Semi Join 是具有 Null-Aware 性质的。不同的是，NAAJ 优化将 NA-EQ 等值条件也作为了 Hash Join 的连接条件，大大加速的了 Join 的计算。

Reviewer 2: ```suggestion
第二个查询 `EXPLAIN SELECT * FROM t WHERE (a, b) NOT IN (select * FROM s);` 中，由于 `t` 表和 `s` 表的 `a`，`b` 列都是 NULLABLE 的，所以 `NOT IN` 子查询所转化的 Anti Semi Join 是具有 Null-Aware 性质的。不同的是，NAAJ 优化将 NA-EQ 条件也作为了 Hash Join 的连接条件，大大加速了 Join 的计算。
```

> 当前 TiDB 仅针对 Anti Semi Join 和 Anti Left Outer Semi Join 实现了 `NULL` 感知。目前仅支持 Hash Join 类型且其 build 表只能固定为右侧表。

Author: 这个是 Hash Join 的一个实现，其接受左右孩子节点的输入，选择其一端构建 hash 表（该侧成为 build 端，该表称为 build 表）。

null-aware hash join 算是一种类型的 Hash Join，但是目前我们无法将其左侧孩子输入构建 hash 表来用
